### PR TITLE
Do not allocate when flushing

### PR DESCRIPTION
**Pull request by pascutto**

### CODE REVIEW

**Comment by samoht** [comment]
This needs to cleaning up - @pascutto will do it one day :-)

**Comment by CraigFe** [comment]
Rebased.

File: src/unix/buffer.ml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+open! Import
```
Comment (samoht) [inline src/unix/buffer.ml:1]: missing headers

File: src/unix/raw.ml [hunk 2]
```diff
@@ -34,9 +34,9 @@ let fsync t = Unix.fsync t.fd
 let close t = Unix.close t.fd
 let fstat t = Unix.fstat t.fd
 
-let unsafe_write t ~off buf =
+let unsafe_write t ~off buf buf_offset length =
```
Comment (icristescu) [inline src/unix/raw.ml:37]: nitpicking, but do you mind renaming `buf_offset` to `buffer_offset` as above?

File: src/unix/raw.ml [hunk 3]
```diff
@@ -85,7 +85,7 @@ module Version = struct
     assert_read ~len n;
     Bytes.unsafe_to_string buf
 
-  let set t v = unsafe_write t ~off v
+  let set t v = unsafe_write t ~off v 0 (String.length v)
```
Comment (icristescu) [inline src/unix/raw.ml:88]: we can replace `String.length v` with `8` here I think

**Review by icristescu — APPROVED** [review]
Looks good, thanks! can you add some memtrace graphs to track the effect of this PR?

**Comment by CraigFe** [comment]
Thanks! Merging.

This is a very significant performance improvement for write-intensive loads. Running benchmarks with:

```bash
; dune exec bench/bench.exe -- --nb-entries 10_000_000 --filter '^replace_random$' --log-size 200_000
```

For this benchmark, this PR reduces total allocations from 35G to 24G, and makes memory usage more stable. Before:

image

After:

image

**Pull request merged**